06_Userspace/06_Resource_Management.md

```diff
@@ -0,0 +1,118 @@
+
+```c
+typedef enum {
+    FILE,
+    MESSAGE_ENDPOINT,
+    SHARED_MEM,
+    // can extend later
+} resource_type_t;
+```
+
+In this example, `FILE` represents a file on the disk, `MESSAGE_ENDPOINT` is used for an IPC message queue, and `SHARED_MEM` for a shared memory region between processes. As the kernel grows this struct can be extended to support more resource types. 
+
+Next, we need a generic representation of a resource inside the kernel. This can be defined by the `resource_t` struct:
+
+```c
+typedef struct {
+    resource_type_t type;
+    void* impl;
+} resource_t;
+```
+
+The `type` field tells the kernel what kind of resource it is, and the `impl` pointer allows the kernel to attach the resource-specific implementation of that resource. For example, a file's `impl` could point to a struct holding the file's offset and inode, or for shared memory, it could point to the physical address of that region. 
+
+## Per Process Resource Table
+With an abstract resource now defined, we can extend our previous definition of a process to include a **resource table**:
+
+```c
+typedef struct {
+size_t pid;
+status_t process_status;
+
+// Other fields
+
+resource_t* resource_table[MAX_RESOURCES];
+} process_t;
+```
+
+Now each process has a resource table that is a map of integers, called _handles_, to the kernel resource objects. A handle is simply an identifier returned by the kernel when opening a resource that is later used by the user to inform what resource the operation should be performed upon. This indirection is important because we do not want to expose any kernel pointers directly to a userspace process. Even if they cannot be used there, passing them could still create security or stability risks. This way, the resource structure is not exposed to userspace. Because of this, the same handle number in different processes can refer to different resources. For example, in Unix, handles `0`, `1`, and `2` refer to stdin, stdout, and stderr, for each process and are called "file descriptors". 
+
+With this, we can also define a supporting function allowing the kernel to fetch a resource by handle:
+
+```c
+resource_t* get_resource(process_t* proc, int handle) {
+
+  // Invalid handle
+  if (handle < 0 || handle >= MAX_HANDLES)
+      return NULL;
+  
+    return proc->table[handle];
+}
+```
+
+You would also want to implement two other functions: `int register_resource(process_t* proc, resource_t* rec)` that finds a free handle and stores the resource in the array and also `int remove_resource(process_t* proc, resource_t* rec)` that then marks that handle usable and frees the memory of that resource on clean up.
```

Comment:
No criticism here, I like the way you've structured this wrt to showing the bare minimum critical code and then describing what might be useful and leaving the reader to figure it out. Feel free to mark this comment as resolved.